gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/access/operations/SortedMap2IndexInsertOperation.java

```diff
@@ -287,18 +290,47 @@ else if (lockTimeout > 0) { // wait before throwing conflict
 
         // TODO: identify the cause of Bug SNAP-2627
         // Till then this is the crude fix
-        if (oldValue instanceof AbstractRegionEntry &&
-            ((AbstractRegionEntry)oldValue).isDestroyedOrRemoved()) {
-          // create a dummy exception
-          Throwable th = GemFireXDUtils.newDuplicateKeyViolation("unique constraint",
-            container.getQualifiedTableName(), "key=" + key.toString()
-            + ", row=" + value, oldValue, null, null);
-          final GemFireCacheImpl cache = Misc.getGemFireCache();
-          final LogWriter logger = cache.getLogger();
-          logger.error("Unique index constraint violation caused due to " +
-            "removed/destroyed entry. The index is corrupted. Cleaning the index", th);
-          skipListMap.remove(key, oldValue);
-          continue;
+        if (oldValue instanceof AbstractRegionEntry) {
+          final AbstractRegionEntry existingRe = (AbstractRegionEntry)oldValue;
+          if (existingRe.isDestroyedOrRemoved()) {
+            Callable<Boolean> deadEntryRemover = new Callable<Boolean>() {
+              @Override
+              public Boolean call() {
+                synchronized (existingRe) {
+                  if (existingRe.isDestroyedOrRemoved()) {
+                    if (!existingRe.isRemovedPhase1()) {
+                      // create a dummy exception
+                      Throwable th = GemFireXDUtils.newDuplicateKeyViolation("unique constraint",
+                          container.getQualifiedTableName(), "key=" + key.toString()
+                              + ", row=" + value, oldValue, null, null);
+                      final GemFireCacheImpl cache = Misc.getGemFireCache();
+                      final LogWriter logger = cache.getLogger();
+                      logger.error("Unique index constraint violation caused due to " +
+                          "removed/destroyed entry. The index is corrupted. Cleaning the index", th);
+
+                      skipListMap.remove(key, oldValue);
+                    }
+                    return true;
+                  }
+
+                }
+                return false;
+              }
+
+            };
+
+            FutureTask<Boolean> outcome = new FutureTask<Boolean>(deadEntryRemover);
+            Thread cleaner = new Thread(outcome);
```

Comment:
Creating a new thread everytime will be quite expensive. Why can't this use one of the pools like DM's function execution or waiting pool?

Reply:
True. I am assuming this happens only if the entry still has token destroyed/removed. but yes, I was not sure which threadpool to use so created on the fly. could you pls add it for me...? so that hotfix can be prepared?